dev-docs/RFCs/vNext/layer-extension-rfc.md

```diff
@@ -0,0 +1,141 @@
+
+Secondly, because of the ad-hoc nature of layer customization, it is difficult to package and reuse an "extension" for other layers. This is apprent in the [kepler.gl repo](https://github.com/uber/kepler.gl/tree/master/src/deckgl-layers) where a custom version of every core layer must be created in order to override a simple default bahavior.
+
+To mitigate this, we can of course add as many customizable functionalities to the base layer as possible. However, it is a major concern that inflating the base layer will sacrefice performance in return. Additional attributes takes CPU cycles to generate and memory to store. Larger shaders take longer to compile. WebGL1 has a restricting limit of 16 attributes per vertex shader, so adding attributes without the user asking for it will make the layer less extensible instead.
+
+
+## Case Studies
+
+### Picking
+
+Components:
+
+* Props: `pickable`, `autoHighlight`, `highlightedObjectIndex`, `highlightColor`
+* Shader module: `picking`
+* Attribute: `instancePickingColors` (generated regardless of the `pickable` setting)
+* Shader injection:
+  - Vertex shader: `picking_setPickingColor(instancePickingColors);`
+  - Fragment shader: `gl_FragColor = picking_filterHighlightColor(gl_FragColor); gl_FragColor = picking_filterPickingColor(gl_FragColor);`
+
+### Data Filter
+
+See [Data Filter RFC](/dev-docs/RFCs/v6.0/data-filter-rfc.md)
+
+Components:
+
+* Props: `getFilterValue`, `filterRange`
+* Shader module: `filter`
+* Attributes: `instanceFilterValue`
+* Shader injection:
+  - Vertex shader: `filter_setVisibility(instanceFilterValue);`
+  - Frament shader: `gl_FragColor = filter_filterColor(gl_FragColor);`
+
+### Brushing
+
+See [ArcBrushingLayer example](https://gist.github.com/Pessimistress/dc2becf3809c67dc443b4dbab1b9a46f#file-index-html-L128)
+
+A generic version of this functionality would need the following components:
+
+* Props: `enableBrush`, `brushRadius`
+* Shader module: `brush`
+* Shader injection:
+  - Vertex shader: `brush_setVisibility(instancePositions);`
+  - Frament shader: `gl_FragColor = brush_filterColor(gl_FragColor);`
+
+
+## Proposal
+
+### Layer Class
+
+Some changes to the base Layer class are needed to make a generic, reusable extension system work:
+
+* New interface that each sublayer must implement. During a layer update, these will be invoked by the base layer lifecycle instead of the current ad-hoc implementation:
+  - `getShaders()`
+  - `createModels(shaders)`
+* New method that an extension can call:
+  - `invalidateModels()`
+* official `bufferLayout` (*API audit needed*) support in attribute management. The `bufferLayout` state is used in PathLayer and PolygonLayer to describe the number of instance for each data object. Currently, `AttributeManager`'s auto-update feature always assumes 1:1 mapping between an instance and a data object. Without supporting variable layout, an extension will have to implement layer-specific updaters for these layers.
```

Comment:
What functions in an "extension" are dependent on the vertex array layouts? It seems like there is a list of things we want to support with the extension system, maybe be good list them and prioritize them etc.